practice/tic-tac-toe/dict_board.py

```diff
@@ -0,0 +1,185 @@
+        ... (2, 2): 'O'
+        ... }
+        >>> a.calc_winner()
+        'O'
+
+        >>> a = DictTTTBoard()
+        >>> a._coord_to_token = {
+        ... (0, 0): 'X', (2, 0): 'O',
+        ... (1, 1): 'X',
+        ... (2, 2): 'X'
+        ... }
+        >>> a.calc_winner()
+        'X'
+        """
+        win_keys = [
+            [self._coord_to_token.get((1, 0), ' '), self._coord_to_token.get((1, 1), ' '),
+             self._coord_to_token.get((1, 2), ' ')],
+
+            [self._coord_to_token.get((0, 1), ' '), self._coord_to_token.get((1, 1), ' '),
+             self._coord_to_token.get((2, 1), ' ')],
+
+            [self._coord_to_token.get((0, 2), ' '), self._coord_to_token.get((1, 1), ' '),
+             self._coord_to_token.get((2, 0), ' ')],
+
+            [self._coord_to_token.get((0, 0), ' '), self._coord_to_token.get((1, 1), ' '),
+             self._coord_to_token.get((2, 2), ' ')],
+
+            [self._coord_to_token.get((0, 0), ' '), self._coord_to_token.get((1, 2), ' '),
+             self._coord_to_token.get((0, 2), ' ')],
+
+            [self._coord_to_token.get((0, 0), ' '), self._coord_to_token.get((0, 1), ' '),
+             self._coord_to_token.get((0, 2), ' ')],
+
+            [self._coord_to_token.get((2, 2), ' '), self._coord_to_token.get((2, 1), ' '),
+             self._coord_to_token.get((2, 0), ' ')],
+
+            [self._coord_to_token.get((2, 2), ' '), self._coord_to_token.get((1, 2), ' '),
+             self._coord_to_token.get((0, 2), ' ')]
+        ]
+
+        if ['X', 'X', 'X'] in win_keys:
+            return 'X'
+        elif ['O', 'O', 'O'] in win_keys:
+            return 'O'
+        else:
+            return None
+
+    def __str__(self):
+        r"""Returns a pretty-printed picture of the board.
+
+        >>> a = DictTTTBoard()
+        >>> a._coord_to_token = {
+        ... (0,0): 'X', (1,0): 'X', (2,0): 'O',
+        ... (0,1): 'O', (1,1): 'O', (2,1): 'X',
+        ... (0,2): 'X', (2,2): 'X'
+        ... }
+        >>> a.__str__()
+        'X|X|O\nO|O|X\nX| |X\n'
+        """
+        cell00 = ' '
+        cell01 = ' '
+        cell02 = ' '
+        cell10 = ' '
+        cell11 = ' '
+        cell12 = ' '
+        cell20 = ' '
+        cell21 = ' '
+        cell22 = ' '
+        self._coord_to_token.items()
+        for pair in list(self._coord_to_token.items()):
+            if pair[0] == (0, 0):
+                cell00 = pair[1]
+            if pair[0] == (0, 1):
+                cell01 = pair[1]
+            if pair[0] == (0, 2):
+                cell02 = pair[1]
+            if pair[0] == (1, 0):
+                cell10 = pair[1]
+            if pair[0] == (1, 1):
+                cell11 = pair[1]
+            if pair[0] == (1, 2):
+                cell12 = pair[1]
+            if pair[0] == (2, 0):
+                cell20 = pair[1]
+            if pair[0] == (2, 1):
+                cell21 = pair[1]
+            if pair[0] == (2, 2):
+                cell22 = pair[1]
```

Comment:
This is another location where the code itself was generated by an "algorithm" and would be best done by the computer itself. Effectively this is making a 3x3 list of lists (but in separately named variables) then manually doing a reverse coordinate lookup to place a token in a slot. Might as well get the computer to do it!

```py
for coords, token in items:
    x, y = coords
    rows[y][x] = token
```

Then it is the list version of the class `__str__()` problem.